recommenders/models/sasrec/util.py

```diff
@@ -0,0 +1,195 @@
+import sys
+import copy
+import random
+import numpy as np
+import tensorflow as tf
+from collections import defaultdict
+from tqdm import tqdm
+
+
+def data_partition(fname):
+    usernum = 0
+    itemnum = 0
+    User = defaultdict(list)
+    user_train = {}
+    user_valid = {}
+    user_test = {}
+    # assume user/item index starting from 1
+    # f = open('data/%s.txt' % fname, 'r')
+    f = open(fname, 'r')
+    for line in f:
+        u, i = line.rstrip().split(' ')
+        u = int(u)
+        i = int(i)
+        usernum = max(u, usernum)
+        itemnum = max(i, itemnum)
+        User[u].append(i)
+
+    for user in User:
+        nfeedback = len(User[user])
+        if nfeedback < 3:
+            user_train[user] = User[user]
+            user_valid[user] = []
+            user_test[user] = []
+        else:
+            user_train[user] = User[user][:-2]
+            user_valid[user] = []
+            user_valid[user].append(User[user][-2])
+            user_test[user] = []
+            user_test[user].append(User[user][-1])
+    return [user_train, user_valid, user_test, usernum, itemnum]
+
+
+def data_partition_with_time(fname, sep=" "):
+    usernum = 0
+    itemnum = 0
+    User = defaultdict(list)
+    Items = set()
+    user_train = {}
+    user_valid = {}
+    user_test = {}
+    # assume user/item index starting from 1
+    f = open('data/%s.txt' % fname, 'r')
+    
+    for line in f:
+        u, i, t = line.rstrip().split(sep)
+        User[u].append((i, t))
+        Items.add(i)
+
+    for user in User.keys():
+        # sort by time
+        items = sorted(User[user], key=lambda x: x[1])
+        # keep only the items
+        items = [x[0] for x in items]
+        User[user] = items
+        nfeedback = len(User[user])
+        if nfeedback == 1:
+            del User[user]
+            continue
+        elif nfeedback < 3:
+            user_train[user] = User[user]
+            user_valid[user] = []
+            user_test[user] = []
+        else:
+            user_train[user] = User[user][:-2]
+            user_valid[user] = []
+            user_valid[user].append(User[user][-2])
+            user_test[user] = []
+            user_test[user].append(User[user][-1])
+    
+    usernum = len(User)
+    itemnum = len(Items)
+    return [user_train, user_valid, user_test, usernum, itemnum]
+
+
+def evaluate(model, dataset, maxlen, num_neg_test):
+    [train, valid, test, usernum, itemnum] = copy.deepcopy(dataset)
+
+    NDCG = 0.0
+    HT = 0.0
+    valid_user = 0.0
+
+    if usernum>10000:
+        users = random.sample(range(1, usernum + 1), 10000)
+    else:
+        users = range(1, usernum + 1)
+    
+    for u in tqdm(users, ncols=70, leave=False, unit='b'):
+
+        if len(train[u]) < 1 or len(test[u]) < 1: continue
+
+        # print(train[u])
```

Comment:
The metrics that are used here, NDCG@10 and Hit@10 are same as what is available in deeprec_utils.py (https://github.com/microsoft/recommenders/blob/main/recommenders/models/deeprec/deeprec_utils.py). Currently I cannot include them since Tf2.x is not supported (_from deeprec_utils import ndcg_score, hit_score_ does not work). Once we migrate to TF 2.x it is easy to invoke those functions.